Sources/BaggageContext/Context.swift

```diff
@@ -0,0 +1,220 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Baggage Context open source project
+//
+// Copyright (c) 2020 Moritz Lang and the Swift Baggage Context project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_exported import Baggage
+import Logging
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: Context Protocol
+
+/// The `ContextProtocol` MAY be adopted by specific "framework contexts" such as e.g. `CoolFramework.Context` in
+/// order to allow users to pass such context directly to libraries accepting any context.
+///
+/// This allows frameworks and library authors to offer APIs which compose more easily.
+/// Please refer to the "Reference Implementation" notes on each of the requirements to know how to implement this protocol correctly.
+public protocol Context {
```

Comment:
Bringing this to the main PR...

What's the thought on making this more tailored to the intended semantics of naming this more like `TracingContext`, `LoggingContext`, or something similar?

A lot of the "confusion" I've had is that `Context` is very generic, and we're writing in several places that this _is not_ a general execution context object and more specifically a "execution metadata" context with baggage metadata being passed around.

Reply:
> A lot of the "confusion" I've had is that Context is very generic, and we're writing in several places that this is not a general execution context object and more specifically a "execution metadata" context with baggage metadata being passed around.
> @Mordil 

I agree ☝️ I'd also like to throw `InstrumentationContext` into the mix, as it will be used for many kinds of instrumentation, not only tracing.

`LoggingContext`, in my opinion, sounds more like something internal to swift-log.

Reply:
Ah yes, I do like `InstrumentationContext` quite a lot

Reply:
CC @ktoso

Reply:
Hey yeah, thanks for the thought - it's a valid concern. I really don't know what a different good name would be other than `BaggageContext`.

- `LoggingContext` is problematic IMHO but if people really love it we could... it does after all concern it self a lot about binding the baggage to logging.
- `InstrumentationContext` pulls in a word into this library that is not from this library -- we can't do that IMHO. We have to be very strict about the words we use and the word instrument is not defined in this library, and never will be really. It just so happens that instrumentation is what uses this, but this is valuable even if instruments didn't exist...
- `TracingContext` is a very specific thing and I really don't want to tie this to tracing per se. It also is the exact name in a few tracing APIs which makes people think it's the same (but it's not)
- Some tracing APIs have called this or similar things `PropagationContext` but actually they're now moved back and just call it `Baggage` (huh)
- Baggage is our "container" so... this really can only be `Context` or `BaggageContext` IMHO.

Since we want to have this "basically everywhere" a least long name is preferable thus `Context`...

If I may, I'd propose either sticking with `Context` or `BaggageContext`.

// cc @tomerd @lukasa @slashmo

Reply:
agree with the sentiment. the closest I can think of is `BaggageContext`

Reply:
It would be a shame to apply the word `Context` too early: if the type is not appropriately general-purpose, we shouldn’t try to steal the name. We may also not want to glue baggage and context too tightly together if we were doing truly general-purpose `Context` stuff.

Reply:
Thanks for the round of comments. We're going in circles here since months as I'm sure you all know... 😉 

--- 

As for the "is this general purpose or not" that's a tricky question as it depends on what one _really_ means by general purpose... 

This is a general purpose _context propagation type_, that's a known concept and the phrase is used a lot in other ecosystems, papers etc. At the same time, a lot of people seem to think about dependency injection stuff when they see "Context" (at least based on my discussions here with OSS folks as well as Swift UI folks and their `Environment` which "looks similar" but is _very_ different in purpose).

So... is this a general purpose context propagation type? Yeah, at least in my book.

---

Attempting to resolve the naming deadlock yet again:

Since this `Context` protocol is purely for:
1. "frameworks to conform their context types to" 
2. be accepted as parameter by libraries / frameworks whenever making outbound calls, and "passing around" towards such calls through a system

It shows up only in:

```swift
// framework
struct/class Lambda.Context: (This)Context { ... }

func call(..., context: (This)Context) { ... } 
// at call site the type is not really "showing up":
// someThing.call(context: someContext)
```

So it mostly shows up in libraries/clients.

_The purpose of the context is to carry around the baggage, and also logger._

This is because of how people want to "pass around the logger" in Swift (unlike Go where `Logger(from: context)` would be natural, or Java/Scala where `Logger.get(class).info(context, ...)` (or implicits or thread locals) would be more natural).

Thus, I propose:

- `protocol BaggageContext { var logger: Logger {get set} var baggage: Baggage {get set} }`
  - the type to conform "framework contexts" to
  - the type to accept in client APIs
  - the module has to depend on `Logging`
  - people can build their contexts to be whatever they want as long as it safely exposes baggage and logger -- this makes our cases happy, and people can go crazy _inside_ their frameworks, but on library boundaries they should accept the BaggageContext type to be as most interoperable as possible.
    - if they don't, people will have to perform mappings etc, but that's a choice of the (framework) API designer
- `struct Baggage`
  - the "storage", the "thing that is carried around"

Yes, it is not "LoggingBaggageContext" (we did have this before, as `LoggingBaggageContextCarrier` and that is quite a mouthful and was not met with all that much move), even though it's primary goal is to bind a logger and context together, but really – it at least explains what it is -- it's some context for purpose of carrying around baggage. And that's what we care about.

I know not everyone loves the word baggage, but it is here to stay: w3c.github.io/baggage "**Propagation format for distributed trace context: Baggage**" whose goal is:

> This specification defines propagation of baggage for events correlation beyond the request identification that is covered by trace context specification. Our goal is to share this with the community so that various tracing and diagnostics products can operate together.

Which is exactly our goal as well.

So... WDYT? 

What would module names be then?

- package `SwiftBaggageContext`
  - module `swift-baggage` / `import SwiftBaggage` with the type `Baggage`
  - module `swift-baggage-context` / `import SwiftBaggageContext` with the type `BaggageContext`
    - tbh I'd like to _export `Baggage` from the baggage context module so it's less of a pain.

The only reason we separate Baggage and BaggageContext is because NIO @Lukasa said they don't want to depend on `Logging` but it would be nice if it depended on `Baggage` for this reason: https://github.com/apple/swift-nio/pull/1574

I'd like to confirm that's still true, and if we actually need baggage / baggage context in separate repositories or not, @lukasa?

Reply:
I am 👍  on your summary

Reply:
I'm good with most of this but prefixing the framework name with "Swift" is an exception to all the other standard server side frameworks. We don't import SwiftNIO, SwiftLogging etc. I know you are looking for a name unique to the symbols inside the framework but is there something else. Maybe use the plural: `import Baggages`, `import BaggageContexts`. 

Also package name should be the same name as the github repo name to avoid having people add the package name in their dependencies.  
```
.package(name: "SwiftBaggageContext", url: "https://github.com/apple/swift-baggage-context.git", from: "1.0.0")
```

Reply:
I agree with the concern about calling it general-purpose, but having it be the context used for propagation. I think the confusion lies in there being different interpretations of "Context". With this library, we only want to expose a context used for propagation, not one (mis-)used e.g. for dependency injection. I'm +1 on `Baggage` and `BaggageContext` as type names as they have prior art in the tracing, which relies on context propagation.

> module `swift-baggage-context` / `import SwiftBaggageContext`
> @ktoso

What I'd suggest for the module name instead of prefixing with `Swift` is to make it extra clear what this is about and call it `ContextPropagation`. The package/repo name could then be `swift-context-propagation`.

Reply:
Thanks for chiming in Adam, the package naming is also important indeed. I would like to avoid "plural names" though. There technically exists precedence for `Swift...` prefixed modules because `SwiftProtobuf` does so, but yes you're right it's not a pattern we follow.

I very much like @slashmo  proposal where the package name is more "functional" rather than just repeating the same words again, we brainstormed this a bit more and I would like to propose the following which I think addresses all concerns:

- `.package("https://github.com/apple/swift-context-propagation.git", from: "1.0.0")`
  - module name: `import ContextPropagation`
  - includes `protocol BaggageContext { Baggage Logger }` and `DefaultContext`, and friends (the log handler etc)
  - depends on `SwiftLog`
- `.package("https://github.com/apple/swift-context-propagation-baggage.git", from: "1.0.0")`
    - libraries which do not want to depend on logging can use this if they still need to carry a baggage somehow (e.g. NIO this way is able to not have to depend on logging, but offer `context.baggage` itself)
  - module name: `import ContextPropagationBaggage`
  - includes `struct Baggage` and the related key types etc
  - zero dependencies


Please note that repository names are subject to input from Apple teams and management etc, so I can't commit to those, but those would seem a very good balanced tradeoff. Not squatting the general term "context" and not looking weird "swift-baggage" which is a bit "what does this repo even do?" for the uninitiated. It also solves our module name != main type requirement mentioned up-thread here.

Reply:
via @Mordil

> Are we still sure there needs to be two repos instead of just one?

Yes, I've quadruple checked this today and consulted with Cory as well.

If we don't, then there's an artificial link between a "i dont want to pull logging" projects that would fetch and enforce a version requirement even though they don't use logging in any module.

Imagining such package:

```
    dependencies: [
        .package(url: the repo),
    ],
    targets: [
        .target(name: "SotoCore", dependencies: [
            .product(name: "Baggage", package: "swift-context"), // "i dont want logging"
        ]),
    ] 
```

we're STILL both fetching and including in version resolution the logging package:

```
-> % swift package update
Fetching https://github.com/ktoso/gsoc-swift-baggage-context.git
Fetching https://github.com/apple/swift-log.git
Cloning https://github.com/apple/swift-log.git
Resolving https://github.com/apple/swift-log.git at 1.4.0
Cloning https://github.com/ktoso/gsoc-swift-baggage-context.git
Resolving https://github.com/ktoso/gsoc-swift-baggage-context.git at simple-is-good-proposal
```

A project which does not use logging at any point (e.g. nio but there can be others), should not force that artificial link between logging and the package. It does not matter that such package does not depend on the module that actually depends on logging for these parts; it helps that logging would not be in the binary result, but for purposes of package management it makes no difference and the logging package still is considered.

Reply:
Ah, that’s a little disappoint that SwiftPM does that. I thought it didn’t consider dependencies not used in the target graph.

Ok then. I’m still on board 💯

Reply:
Let's get those ✅ rolling in 🏁 (I hope?)